src/GafferUI/PlugAdder.cpp

```diff
@@ -0,0 +1,354 @@
+
+IE_CORE_DEFINERUNTIMETYPED( PlugAdder );
+
+PlugAdder::PlugAdder( Gaffer::NodePtr node, StandardNodeGadget::Edge edge )
+	:	m_node( node ), m_edge( edge ), m_dragging( false )
+{
+	node->childAddedSignal().connect( boost::bind( &PlugAdder::childAdded, this ) );
+	node->childRemovedSignal().connect( boost::bind( &PlugAdder::childRemoved, this ) );
+
+	enterSignal().connect( boost::bind( &PlugAdder::enter, this, ::_1, ::_2 ) );
+	leaveSignal().connect( boost::bind( &PlugAdder::leave, this, ::_1, ::_2 ) );
+	buttonPressSignal().connect( boost::bind( &PlugAdder::buttonPress, this, ::_1,  ::_2 ) );
+	dragBeginSignal().connect( boost::bind( &PlugAdder::dragBegin, this, ::_1, ::_2 ) );
+	dragEnterSignal().connect( boost::bind( &PlugAdder::dragEnter, this, ::_2 ) );
+	dragMoveSignal().connect( boost::bind( &PlugAdder::dragMove, this, ::_1, ::_2 ) );
+	dragLeaveSignal().connect( boost::bind( &PlugAdder::dragLeave, this, ::_2 ) );
+	dropSignal().connect( boost::bind( &PlugAdder::drop, this, ::_2 ) );
+	dragEndSignal().connect( boost::bind( &PlugAdder::dragEnd, this, ::_2 ) );
+
+	updateVisibility();
+}
+
+PlugAdder::~PlugAdder()
+{
+}
+
+Imath::Box3f PlugAdder::bound() const
+{
+	return Box3f( V3f( -0.5f, -0.5f, 0.0f ), V3f( 0.5f, 0.5f, 0.0f ) );
+}
+
+void PlugAdder::updateDragEndPoint( const Imath::V3f position, const Imath::V3f &tangent )
+{
+	m_dragPosition = position;
+	m_dragTangent = tangent;
+	m_dragging = true;
+	requestRender();
+}
+
+void PlugAdder::doRender( const Style *style ) const
+{
+	if( m_dragging )
+	{
+		if( !IECoreGL::Selector::currentSelector() )
+		{
+			V3f srcTangent( 0.0f, 0.0f, 0.0f );
+			style->renderConnection( V3f( 0 ), srcTangent, m_dragPosition, m_dragTangent, Style::HighlightedState );
+		}
+	}
+
+	float radius = 0.75f;
+	Style::State state = Style::NormalState;
+	if( getHighlighted() )
+	{
+		radius = 1.25f;
+		state = Style::HighlightedState;
+	}
+	style->renderImage( Box2f( V2f( -radius ), V2f( radius ) ), texture( state ) );
+}
+
+void PlugAdder::addPlug( Gaffer::Plug *connectionEndPoint )
+{
+	UndoContext undoContext( m_node->ancestor<ScriptNode>() );
+
+	if( SwitchComputeNode *switchNode = runTimeCast<SwitchComputeNode>( m_node.get() ) )
+	{
+		switchNode->setup( connectionEndPoint );
+		ArrayPlug *inPlug = switchNode->getChild<ArrayPlug>( "in" );
+		Plug *outPlug = switchNode->getChild<Plug>( "out" );
+
+		StandardNodeGadget::Edge inEdge = StandardNodeGadget::InvalidEdge;
+		if( connectionEndPoint->direction() == Plug::Out )
+		{
+			inPlug->getChild<Plug>( 0 )->setInput( connectionEndPoint );
+			inEdge = m_edge;
+		}
+		else
+		{
+			connectionEndPoint->setInput( switchNode->getChild<Plug>( "out" ) );
+			inEdge = oppositeEdge( m_edge );
+		}
+
+		updateMetadata( inPlug, "nodeGadget:nodulePosition", edgeName( inEdge ) );
+		updateMetadata( inPlug, "compoundNodule:orientation", edgeOrientation( inEdge ) );
+		updateMetadata( outPlug, "nodeGadget:nodulePosition", edgeName( oppositeEdge( inEdge ) ) );
```

Comment:
So at IE we do specify color (both nodule and connection) by plug type: https://github.com/ImageEngine/gafferExtensions/blob/master/config/lightingTools/colorScheme.py#L20-L45

And thats resulting in "what I'd expect" for scene/image/filter/task:

![switchcolors](https://cloud.githubusercontent.com/assets/3979682/20906897/4dcd6046-baff-11e6-9774-1bae1a5de839.png)

But something "funky" for color and float (despite being specified in that same section of the colorScheme config). I guess those colors are being overridden by this https://github.com/johnhaddon/gaffer/blob/12b2dc4f495779ac9cf0710dd1d50965a6ce1ca8/startup/gui/nodeGraph.py#L84-L88, but wouldn't that apply to the original shader as well?

![switchcolorsfunky](https://cloud.githubusercontent.com/assets/3979682/20906946/b1373f12-baff-11e6-8efb-1d27b490a33e.png)